pkg/validation/validation.go

```diff
@@ -0,0 +1,83 @@
+package validation
+
+import (
+	"fmt"
+	"os"
+
+	"github.com/ghodss/yaml"
+	"github.com/xeipuuv/gojsonschema"
+)
+
+// https://github.com/xeipuuv/gojsonschema#formats
+type ValidFormat struct{}
+
+// IsFormat always returns true and meets the
+// gojsonschema.FormatChecker interface
+func (f ValidFormat) IsFormat(input interface{}) bool {
+	return true
+}
+
+// Based on https://stackoverflow.com/questions/40737122/convert-yaml-to-json-without-struct-golang
+// We unmarshal yaml into a value of type interface{},
+// go through the result recursively, and convert each encountered
+// map[interface{}]interface{} to a map[string]interface{} value
+// required to marshall to JSON.
+// Reference: https://github.com/garethr/kubeval/blob/master/kubeval/utils.go#L8
+func convertToStringKeys(i interface{}) interface{} {
+	switch x := i.(type) {
+	case map[interface{}]interface{}:
+		m2 := map[string]interface{}{}
+		for k, v := range x {
+			m2[k.(string)] = convertToStringKeys(v)
+		}
+		return m2
+	case []interface{}:
+		for i, v := range x {
+			x[i] = convertToStringKeys(v)
+		}
+	}
+	return i
+}
+
+// validate function will validate input kedgefile against JSON schema provided in pkg/validation
+func Validate(p []byte) {
+	var speco interface{}
+	err := yaml.Unmarshal(p, &speco)
+	if err != nil {
+		fmt.Printf("Error with Unmarhsalling")
+	}
+	body := convertToStringKeys(speco)
+	s := body.(map[string]interface{})
+	loader := gojsonschema.NewGoLoader(body)
+	if s["controller"] == nil {
+		s["controller"] = "deployment"
+	}
+
+	var schema gojsonschema.JSONLoader
+	// Depend on type of controller, we are fetching jsonschema stored in pkg/validation
+	if s["controller"] == "deployment" {
```

Comment:
this doesn't look good :-(, we will have to always think about adding new if statement here if we add a new controller.
This should be handled in the more generic way. 

Could this be all validation handled in `Validate`  function for each controller object?

like here:
https://github.com/kedgeproject/kedge/blob/b8a42790cd2164c1f0ce1109b0760c383ca660b3/pkg/spec/deployment.go#L42


ping @containscafeine

Reply:
if we have to keep this (but I would rather avoid this) that this should be converted to `switch` with `default` statement that throws error, so we at least have a chance to catch that some controller is missing here.